### PR TITLE
Create spotlight search component

### PR DESCRIPTION
**Pull request by irahov**
The Spotlight Search component is inspired by the Spotlight Search functionality in Mac OSX (cmd+space) and the Search Everywhere in IntelliJ (shift+shift)

This is the basic implementation consisting of:
- angular component
- search service
- search provider interface
- search results interface

The commit also includes
- documentation
- full feature live doc example
- unit tests

Signed-off-by: Ivo Rahov <irahov@vmware.com>

### CODE REVIEW

File: projects/components/src/spotlight-search/spotlight-search.component.ts [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+        if (open) {
+            this.searchSections = this.searchService
+                .getRegisteredProviders()
+                .map(data => ({ ...data, results: [], isLoading: true }));
+            this.doSearch();
+
+            setTimeout(() => {
+                this.searchInput.nativeElement.focus();
+                this.searchInput.nativeElement.select();
+            }, 10);
```
Comment (juanmendes) [inline projects/components/src/spotlight-search/spotlight-search.component.ts:262]: Why 10? Shouldn't 0 be enough?

File: projects/components/src/spotlight-search/spotlight-search.component.html [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+                class="clr-input"
+                [placeholder]="placeholder || ''"
+                [(ngModel)]="searchCriteria"
+            />
+        </div>
+        <div class="search-result-container">
+            <div *ngFor="let searchSection of searchSections" class="search-result-section">
+                <div
+                    *ngIf="
+                        searchSections.length > 1 &&
```
Comment (juanmendes) [inline projects/components/src/spotlight-search/spotlight-search.component.html:23]: Please avoid long expressions in templates https://angular.io/guide/template-syntax#simplicity Create a simple method

File: projects/components/src/spotlight-search/spotlight-search.component.html [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+                        (searchSection.isLoading || searchSection.results.length > 0 || showEmptySection)
```
Comment (juanmendes) [inline projects/components/src/spotlight-search/spotlight-search.component.html:24]: This could still show an empty header. I think we need to not show any sections without results. If If it turns out to be without any sections, it needs a placeholder

File: projects/examples/src/components/spotlight-search/spotlight-search-example.component.ts [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+            0
+        );
+    }
+
+    private resetMousetrap: () => void = () => {};
+
+    ngOnInit(): void {
+        mousetrap.bind(this.kbdShortcut, () => {
+            this.spotlightOpen = true;
+            this.changeDetectorRef.detectChanges();
```
Comment (juanmendes) [inline projects/examples/src/components/spotlight-search/spotlight-search-example.component.ts:52]: Why is this needed? If it really is, please explain

File: projects/components/src/spotlight-search/spotlight-search.component.spec.ts [hunk 5]
```diff
@@ -0,0 +1,607 @@
...
+        return this.findElement('.search-input-container input').nativeElement;
+    }
+
+    public get searchInputValue(): string {
+        return this.searchInputElement.value;
+    }
+
+    public set searchInputValue(val: string) {
+        this.searchInputElement.value = val;
+        this.searchInputElement.dispatchEvent(new Event('input'));
```
Comment (juanmendes) [inline projects/components/src/spotlight-search/spotlight-search.component.spec.ts:556]: Can we move the `dispatchEvent(...)` to the base class?

```ts
protected setInputText(el, text ) {


}
```

File: projects/examples/src/components/spotlight-search/spotlight-search-example.component.ts (hunk 4 above)
Comment (irahov) [reply projects/examples/src/components/spotlight-search/spotlight-search-example.component.ts:52]: Since this happens outside angular zone we need to notify angular manually
otherwise, the user should wait for any operation that will trigger angular change detection (angular zone event)
